test/extensions/filters/http/common/jwks_fetcher_test.cc

```diff
@@ -167,9 +179,166 @@ TEST_F(JwksFetcherTest, TestSpanPassedDown) {
           }));
 
   // Act
-  fetcher->fetch(uri_, parent_span_, receiver);
+  fetcher->fetch(parent_span_, receiver);
+}
+
+const std::string jwks_text = R"(
+{
+  "keys": [
+        {
+          "kty": "RSA",
+          "alg": "RS256",
+          "use": "sig",
+          "kid": "62a93512c9ee4c7f8067b5a216dade2763d32a47",
+          "n": "0YWnm_eplO9BFtXszMRQNL5UtZ8HJdTH2jK7vjs4XdLkPW7YBkkm_2xNgcaVpkW0VT2l4mU3KftR-6s3Oa5Rnz5BrWEUkCTVVolR7VYksfqIB2I_x5yZHdOiomMTcm3DheUUCgbJRv5OKRnNqszA4xHn3tA3Ry8VO3X7BgKZYAUh9fyZTFLlkeAh0-bLK5zvqCmKW5QgDIXSxUTJxPjZCgfx1vmAfGqaJb-nvmrORXQ6L284c73DUL7mnt6wj3H6tVqPKA27j56N0TB1Hfx4ja6Slr8S4EB3F1luYhATa1PKUSH8mYDW11HolzZmTQpRoLV8ZoHbHEaTfqX_aYahIw",
+          "e": "AQAB"
+        },
+        {
+          "kty": "RSA",
+          "alg": "RS256",
+          "use": "sig",
+          "kid": "b3319a147514df7ee5e4bcdee51350cc890cc89e",
+          "n": "qDi7Tx4DhNvPQsl1ofxxc2ePQFcs-L0mXYo6TGS64CY_2WmOtvYlcLNZjhuddZVV2X88m0MfwaSA16wE-RiKM9hqo5EY8BPXj57CMiYAyiHuQPp1yayjMgoE1P2jvp4eqF-BTillGJt5W5RuXti9uqfMtCQdagB8EC3MNRuU_KdeLgBy3lS3oo4LOYd-74kRBVZbk2wnmmb7IhP9OoLc1-7-9qU1uhpDxmE6JwBau0mDSwMnYDS4G_ML17dC-ZDtLd1i24STUw39KH0pcSdfFbL2NtEZdNeam1DDdk0iUtJSPZliUHJBI_pj8M-2Mn_oA8jBuI8YKwBqYkZCN1I95Q",
+          "e": "AQAB"
+       }
+    ]
+ }
+ )";
+
+TEST_F(JwksFetcherTest, TestRetryOnceThenSucceed) {
+  const char retry[] = R"(
+      http_uri:
+        uri: https://pubkey_server/pubkey_path
+        cluster: pubkey_cluster
+
+      retry_policy:
+        retry_back_off:
+          base_interval: 0.001s
+          max_interval: 0.1s
+        num_retries: 3
+   )";
+
+  // Setup
+  setupFetcher(retry);
+  MockUpstream mock_pubkey(mock_factory_ctx_.cluster_manager_, "200", publicKey);
+  MockJwksReceiver receiver;
+
+  Event::MockDispatcher dispatcher;
+  Event::MockTimer* retry_timer;
+  Event::TimerCb retry_timer_cb;
+
+  std::unique_ptr<JwksFetcher> fetcher(
+      JwksFetcher::create(mock_factory_ctx_.cluster_manager_, remote_jwks_, dispatcher));
+
+  EXPECT_CALL(dispatcher, createTimer_(_))
+      .WillRepeatedly(Invoke([&retry_timer, &retry_timer_cb](Event::TimerCb timer_cb) {
+        retry_timer = new Event::MockTimer();
+        retry_timer_cb = timer_cb;
+        EXPECT_CALL(*retry_timer, enableTimer(_, _)).WillRepeatedly(Invoke([&retry_timer_cb]() {
+          retry_timer_cb();
+        }));
+        return retry_timer;
+      }));
+
+  Http::MockAsyncClientRequest request(
+      &(mock_factory_ctx_.cluster_manager_.thread_local_cluster_.async_client_));
+
+  int numFailures = 0;
+  int numSuccesses = 0;
+
+  EXPECT_CALL(mock_factory_ctx_.cluster_manager_.thread_local_cluster_.async_client_,
+              send_(_, _, _))
```

Comment:
Can we re-use MockUpstream by storing "cb" into a vector in the class member?  The the testing code can call them any way they want,  this make the testing code more read-able.

Reply:
Still investigating how to make these "retrying test series" readable in this file and the async test one. 
It is, as far as I know, the only outstanding comment.

What's now in there is functional though.

Reply:
My idea of modifying MockUpstream  or another mock class

```
class MockUpstream {
public:
  MockUpstream(Upstream::MockClusterManager& mock_cm,
                           Http::AsyncClient::FailureReason reason)
    : request_(&mock_cm.thread_local_cluster_.async_client_) {
   ON_CALL(mock_cm.thread_local_cluster_.async_client_, send_(testing::_, testing::_, testing::_))
      .WillByDefault(testing::Invoke(
          [this, reason](Http::RequestMessagePtr&, Http::AsyncClient::Callbacks& cb,
                         const Http::AsyncClient::RequestOptions&) -> Http::AsyncClient::Request* {
            asyn_cb_vector_.push_back(&cb);   // just save it to the vector
            return &request_;
          }));
}

private:
  Http::MockAsyncClientRequest request_;
  std::vector<Http::AsyncClient::Callbacks*> asyn_cb_vector_;
};
```

In the testing code,  
you can verify it is called by:
```
   EXPECT_EQ(mock_upstream.asyn_cb_vector_.size(),   1);
```
and you can call it anyway you want,  failure or success.   like
```
   mock_upstream.asyn_cb_vector_[0]->OnFailure(...)
```

Reply:
implemented in a class named MockedDelayedActionUpstream for lack of better name, just so that it did not force updates to a great many tests.